### PR TITLE
ledger-tool: Refactor accounts subcommand output code

### PR DESCRIPTION
**Pull request by steviez**
#### Problem
The implementation of the `accounts` subcommand currently streams account information as each account is handled. This prevents memory pressure from collecting all of the information in a `Vec`, and is more efficient by doing the I/O as we go instead of waiting until the end. But, this implementation is a bit cluttered.

#### Summary of Changes
This PR does the following:
- Shift the logic for outputting accounts over to `output.rs`
- Add an object `AccountScanner` to handle
- Allows running with `--output json` to output the summary information as well while still being a valid json object
    - This is accomplished by putting the list of accounts in an array for `accounts` field, and then adding a `summary` field

This change lays the groundwork for cleanly adding modes where the output is limited to individual accounts or accounts owned by a specific program

### CODE REVIEW

**Comment by steviez** [comment]
https://github.com/solana-labs/solana/pull/34925 is the continuation of this PR that actually adds some new functionality and is built on top of this PR's branch. I'm open to keeping the PR's separate (ie getting this one pushed and then rebasing the next one) or handling them separately.

File: ledger-tool/src/output.rs [hunk 1]
```diff
@@ -551,6 +561,92 @@ pub fn output_sorted_program_ids(program_ids: HashMap<Pubkey, u64>) {
...
+    total_accounts_stats: Rc<RefCell<TotalAccountsStats>>,
+}
+
+impl AccountsOutputStreamer {
+    pub fn new(
+        bank: Arc<Bank>,
+        include_sysvars: bool,
+        include_account_contents: bool,
+        include_account_data: bool,
+        account_data_encoding: UiAccountEncoding,
```
Comment (t-nelson) [inline ledger-tool/src/output.rs:582]: these are probably candidates for a config struct

File: ledger-tool/src/output.rs [hunk 2]
```diff
@@ -551,6 +561,92 @@ pub fn output_sorted_program_ids(program_ids: HashMap<Pubkey, u64>) {
...
+            .serialize_field("accounts", &self.account_scanner)
+            .unwrap();
+        struct_serializer
+            .serialize_field("summary", &*self.total_accounts_stats.borrow())
+            .unwrap();
+        SerializeStruct::end(struct_serializer).unwrap();
+    }
+}
+
+pub struct AccountsScanner {
```
Comment (t-nelson) [inline ledger-tool/src/output.rs:624]: doesn't need to be public i don't think?

File: ledger-tool/src/output.rs [hunk 3]
```diff
@@ -626,19 +626,25 @@ pub struct AccountsScanner {
     account_data_encoding: UiAccountEncoding,
 }
 
+impl AccountsScanner {
```
Comment (t-nelson) [inline ledger-tool/src/output.rs:633]: why the separate impl context?

File: ledger-tool/src/output.rs [hunk 4]
```diff
@@ -594,21 +599,17 @@ impl AccountsOutputStreamer {
...
     pub fn output(&self) {
         match self.output_format {
-            OutputFormat::Json | OutputFormat::JsonCompact => self.output_json(),
+            OutputFormat::Json | OutputFormat::JsonCompact => self.output_json().unwrap(),
             _ => self.print(),
         }
     }
 
-    fn output_json(&self) {
+    fn output_json(&self) -> Result<(), serde_json::Error> {
```
Comment (t-nelson) [inline ledger-tool/src/output.rs]: prefer avoiding dependency/external types in our APIs

**Review by t-nelson — COMMENTED** [review]
couple api nits since we're going to all the trouble of cleaning this up

File: ledger-tool/src/output.rs (hunk 1 above)
Comment (steviez) [reply ledger-tool/src/output.rs]: Fair enough - done in https://github.com/solana-labs/solana/pull/34915/commits/47fb5850b37e1b21f6a6e27f2fcd5b3e5eaaa430

**Review by steviez — COMMENTED** [review]
Review feedback addressed in separate commits; force push was to rebase to more recent tip of master

File: ledger-tool/src/output.rs (hunk 4 above)
Comment (steviez) [reply ledger-tool/src/output.rs]: Fair enough; ended up mapping these these to a `String` in https://github.com/solana-labs/solana/pull/34915/commits/a655c45a74619817069a9c21d495a80e9f4104a5. Hypothetically, I could make a subfunction for json, use `?` within it and then do a single `.unwrap()` or `.map_err()` on it. There are only 4 `.map_err()`'s so doesn't seem awful as I just pushed it in

**Review by CriesofCarrots — DISMISSED** [review]
Thanks for adding that config struct 😅 
This works for me, but would be good to get sign-off from @t-nelson

File: ledger-tool/src/output.rs (hunk 3 above)
Comment (t-nelson) [reply ledger-tool/src/output.rs:633]: yeah the methods fine i was just wondering why it wasn't put under the `impl AccountsScanner` block at ln635. afaik the compiler will collapse them, but a little jarring in the code

**Review by t-nelson — DISMISSED** [review]
optional nit

File: ledger-tool/src/output.rs (hunk 3 above)
Comment (steviez) [reply ledger-tool/src/output.rs:633]: D'oh; I completely missed your point the first time; I think I just refactored it enough times that I had blinders on what you were calling out. Done in https://github.com/solana-labs/solana/pull/34915/commits/00b3088bbb79550c8ad27455f33dd56fc48fcf05

**Review by t-nelson — APPROVED** [review]

**Review by CriesofCarrots — APPROVED** [review]

**Pull request merged**